File: rd-net/RdCore/Util/BitSlice.cs
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using JetBrains.Annotations;
using JetBrains.Diagnostics;

namespace JetBrains.Util.Util
{
  public class BitSlice
  {
    [PublicAPI] public int LoBit;
    [PublicAPI] public int BitCount;

    [PublicAPI] public int HiBit 
    {
      [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
      get => LoBit + BitCount - 1;
    }

    [PublicAPI] public int Mask; //frequently used field, so it can't should be stored rather than calculated 

    public BitSlice(int loBit, int bitCount)
    {
      LoBit = loBit;
      BitCount = bitCount;
      Mask = (1 << BitCount) - 1;
    }

    
    #region Assertions

    [AssertionMethod]
    private void AssertSliceFitsHostType<T>()
    {
      var type = typeof(T);      
      var maxBit = type == typeof(int) ? 31
        : type == typeof(long) ? 64
        : 0;
      
      Assertion.Assert(maxBit > 0, "Unsupported host type: {0}", type);      
      Assertion.Assert(HiBit <= maxBit, "{0} doesn't fit into host type {1}; must be inside [0, {2}]", this, type, maxBit);
    }

    [AssertionMethod]
    private void AssertValueFitsSlice(int value)
    {
      Assertion.Assert(value >= 0, "[{0}] must be >= 0; actual: {1}", nameof(value), value);
      Assertion.Assert(value <= Mask, "[{0}] must be <= {1} to fit {2}; actual: {3}", nameof(value), Mask, this, value);
    }
    
    #endregion


    #region For inheritance
        
    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
    protected int GetRaw(int host)
    {
//      AssertSliceFitsHostType<int>();
      
      return (host >> LoBit) & Mask;
    }

    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
    protected int UpdatedRaw(int host, int value)
    {
//      AssertSliceFitsHostType<int>();
//      AssertValueFitsSlice(value);
      
      return (host & ((Mask << LoBit) ^ -1)) | (value << LoBit);
    }
    
    #endregion


    #region Static API

    private static int NextSliceLoBit([CanBeNull] BitSlice slice) => slice?.HiBit + 1 ?? 0; 
    
    public static IntBitSlice Int(int bitCount, [CanBeNull] BitSlice previousSlice = null)
    {
      return new IntBitSlice(NextSliceLoBit(previousSlice), bitCount);
    }

    public static BoolBitSlice Bool([CanBeNull] BitSlice previousSlice = null)
    {
      return new BoolBitSlice(NextSliceLoBit(previousSlice), 1);
    }
    
    public static EnumBitSlice<T> Enum<T>([CanBeNull] BitSlice previousSlice = null) where T:struct
    {
      return new EnumBitSlice<T>(NextSliceLoBit(previousSlice));
    }
    

    #endregion
    
    
    public override string ToString() => $"BitSlice[{LoBit}, {HiBit}]";
  }

  
  public abstract class BitSlice<T> : BitSlice
  {
    protected BitSlice(int loBit, int bitCount) : base(loBit, bitCount) {}

    public abstract T this[int host] { get; }
    public abstract int Updated(int host, T value);

    public void InterlockedUpdate(ref int host, T value)
    {
      while (true)
      {
        var s = host;
        if (Interlocked.CompareExchange(ref host, Updated(host, value), s) == s)
        {
          break;
        }
      }
    }
    
    public override string ToString() => $"TypedBitSlice<{typeof(T).Name}>[{LoBit}, {HiBit}]";
  }

  public class IntBitSlice : BitSlice<int>
  {
    public IntBitSlice(int loBit, int bitCount) : base(loBit, bitCount){}
        
    public override int this[int host]
    {
      [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
      get => GetRaw(host);
    }

    public override int Updated(int host, int value) => UpdatedRaw(host, value);
  }
  
  public class BoolBitSlice : BitSlice<bool>
  {
    public BoolBitSlice(int loBit, int bitCount) : base(loBit, bitCount){}
    public override bool this[int host]
    {
      [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]      
      get => GetRaw(host) != 0;
    }

    public override int Updated(int host, bool value) => UpdatedRaw(host, value ? 1 : 0);
  }
  
  public class EnumBitSlice<T> : BitSlice<T> where T:struct
  {
    private static int CalculateBitCount()
    {
      var type = typeof(T);
#if !NETSTANDARD
      Assertion.Require(type.IsEnum, "Type must be enum, actual type: {0}", type);
#endif

      var values = ((T[]) System.Enum.GetValues(typeof(T))).Select(CastTo<long>.From).ToList();
      
      Assertion.Require(values.Count > 0, "Bit slice for enum with no value is meaningless");
      Assertion.Require(values.All(v => v >= 0), "Enums with negative values or values greater than long.MaxValue are unsupported");

      var count = values.Max() + 1;            
      var res = BitHacks.Log2Ceil(count);
      if (res == 0) res++; //for enums with one value

      return res;
    }
    
    public EnumBitSlice(int loBit) : base(loBit, CalculateBitCount()){}
    public override T this[int host]
    {
      [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
      get => CastTo<T>.From(GetRaw(host));
    }

    public override int Updated(int host, T value) => UpdatedRaw(host, CastTo<int>.From(value));
  }

}

File: rd-kt/rd-core/src/main/kotlin/com/jetbrains/rd/util/BitSlice.kt
package com.jetbrains.rd.util

open class BitSlice(val lowBit: Int, val bitCount: Int) {
    init {
        require(lowBit >= 0)  { "[lowBit] must be >= 0, actual: '$lowBit'"}
        require(bitCount > 0) { "[bitCount] must be > 0, actual '$bitCount'"}
    }

    val highBit: Int get() = lowBit + bitCount - 1
    private val mask : Int get() = (1 shl bitCount) - 1

    abstract class TypedBitSlice<T>(lowBit: Int, bitCount: Int) : BitSlice(lowBit, bitCount) {
        abstract operator fun get(host: Int) : T
        abstract fun updated(host: Int, value: T) : Int
    }

    private class IntBitSlice(lowBit: Int, bitCount: Int) : TypedBitSlice<Int>(lowBit, bitCount) {
        override fun get(host: Int) = getRaw(host)
        override fun updated(host: Int, value: Int) = updatedRaw(host, value)
    }

    companion object {
        fun int(bitCount: Int, prev: BitSlice? = null) : TypedBitSlice<Int> = IntBitSlice(prev.nextSliceLow(), bitCount)

        fun bool(prev: BitSlice? = null) : TypedBitSlice<Boolean> = object : TypedBitSlice<Boolean>(prev.nextSliceLow(), 1) {
            override fun get(host: Int) = getRaw(host) != 0
            override fun updated(host: Int, value: Boolean) = updatedRaw(host, if (value) 1 else 0)
        }

        inline fun <reified T:Enum<T>> enum(prev: BitSlice? = null) : TypedBitSlice<T> {
            val values = enumValues<T>()
            require(values.size >= 2) { "Bit slice for enums with $values values is meaningless" }
            return object : TypedBitSlice<T>(prev.nextSliceLow(), log2ceil(values.size)) {
                override fun get(host: Int): T = values[getRaw(host)]
                override fun updated(host: Int, value: T) = updatedRaw(host, value.ordinal)
            }
        }
    }


    private fun requireSliceFitsIntType() {
        val maxBit = 31
        require(highBit <= maxBit) { "$this doesn't fit into host type 'Int'; must be inside [0, $maxBit]"}
    }

    private fun requireValueFitsSlice(value: Int) {
        require(value >= 0) { "[value] must be >= 0; actual: '$value'" }
        require(value <= mask) {"[value] must be <= $mask to fit $this; actual '$value'"}
    }


    //GET and SET
    fun getRaw(host: Int) : Int {
        requireSliceFitsIntType()

        return (host ushr lowBit) and mask
    }


    fun updatedRaw(host: Int, value: Int) : Int {
        requireSliceFitsIntType()
        requireValueFitsSlice(value)

        return (host and  ((mask shl lowBit) xor -1)) or (value shl lowBit)
    }


    override fun toString(): String {
        return "BitSlice[$lowBit, $highBit]"
    }
}

fun BitSlice?.nextSliceLow() = this?.highBit?.plus(1) ?: 0

operator fun <T> BitSlice.TypedBitSlice<T>.get(host: AtomicInteger) = get(host.get())